relay-kafka/src/producer/utils.rs

```diff
@@ -4,20 +4,56 @@ use rdkafka::producer::{DeliveryResult, ProducerContext};
 use rdkafka::{ClientContext, Message};
 use relay_statsd::metric;
 
-use crate::statsd::KafkaCounters;
+use crate::statsd::{KafkaCounters, KafkaGauges};
 
-/// Kafka producer context that logs producer errors.
+/// Kafka client and producer context that logs statistics and producer errors.
 #[derive(Debug)]
-pub struct CaptureErrorContext;
+pub struct Context;
 
-impl ClientContext for CaptureErrorContext {}
+impl ClientContext for Context {
+    /// Report client statistics as statsd metrics.
+    ///
+    /// This method is only called if `statistics.interval.ms` is configured.
+    fn stats(&self, statistics: rdkafka::Statistics) {
+        relay_statsd::metric!(gauge(KafkaGauges::MessageCount) = statistics.msg_cnt);
+        relay_statsd::metric!(gauge(KafkaGauges::MessageCountMax) = statistics.msg_max);
+        relay_statsd::metric!(gauge(KafkaGauges::MessageSize) = statistics.msg_size);
+        relay_statsd::metric!(gauge(KafkaGauges::MessageSizeMax) = statistics.msg_size_max);
 
-impl ProducerContext for CaptureErrorContext {
+        for broker in statistics.brokers.values() {
+            relay_statsd::metric!(
+                gauge(KafkaGauges::OutboundBufferRequests) = broker.outbuf_cnt as u64,
+                broker_name = &broker.name
+            );
+            relay_statsd::metric!(
+                gauge(KafkaGauges::OutboundBufferMessages) = broker.outbuf_msg_cnt as u64,
+                broker_name = &broker.name
+            );
+            if let Some(connects) = broker.connects {
+                relay_statsd::metric!(
+                    gauge(KafkaGauges::Connects) = connects as u64,
+                    broker_name = &broker.name
+                );
+            }
+            if let Some(disconnects) = broker.disconnects {
+                relay_statsd::metric!(
+                    gauge(KafkaGauges::Disconnects) = disconnects as u64,
+                    broker_name = &broker.name
+                );
+            }
+        }
+    }
+}
+
+impl ProducerContext for Context {
     type DeliveryOpaque = ();
 
     /// This method is called after attempting to send a message to Kafka.
     /// It's called asynchronously for every message, so we want to handle errors explicitly here.
     fn delivery(&self, result: &DeliveryResult, _delivery_opaque: Self::DeliveryOpaque) {
+        // TODO: any `Accepted` outcomes (e.g. spans) should be logged here instead of on the caller side,
+        // such that we do not over-report in the error case.
```

Comment:
Do we have a ticket for this?

Reply:
We do now: https://github.com/getsentry/relay/issues/3318